File: backend/app/routers/payroll.py
from datetime import datetime
from typing import List

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import extract
from sqlalchemy.orm import Session

from ..auth import get_current_admin_user, get_current_user
from ..database import get_db
from ..models import MonthlyClosing, PayrollRecord, User, WorkRecord
from ..schemas import MonthlySummary, PayrollAdjustment, PayrollRecordResponse
from ..services.calculation import aggregate_monthly_hours
from ..services.payroll import calculate_payroll

router = APIRouter()


def _get_year_month_parts(year_month: str):
    """Split YYYY-MM into (int year, int month), raising 400 on bad format."""
    try:
        year, month = year_month.split("-")
        return int(year), int(month)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="year_month は YYYY-MM 形式で指定してください",
        )


@router.get("/{year_month}/summary", response_model=List[MonthlySummary])
def get_monthly_summary(
    year_month: str,
    db: Session = Depends(get_db),
    _admin=Depends(get_current_admin_user),
):
    """Return a monthly summary with hours per employee (admin only)."""
    year, month = _get_year_month_parts(year_month)

    users = db.query(User).filter(User.is_active == True).all()
    summaries = []

    for user in users:
        records = (
            db.query(WorkRecord)
            .filter(
                WorkRecord.user_id == user.id,
                extract("year", WorkRecord.work_date) == year,
                extract("month", WorkRecord.work_date) == month,
            )
            .all()
        )

        hours = aggregate_monthly_hours(records)

        closing = (
            db.query(MonthlyClosing)
            .filter(
                MonthlyClosing.user_id == user.id,
                MonthlyClosing.year_month == year_month,
            )
            .first()
        )
        closing_status = closing.status if closing else "open"

        summaries.append(
            MonthlySummary(
                user_id=user.id,
                user_name=user.name,
                year_month=year_month,
                total_hours=hours["total_hours"],
                regular_hours=hours["regular_hours"],
                overtime_hours=hours["overtime_hours"],
                late_night_hours=hours["late_night_hours"],
                holiday_hours=hours["holiday_hours"],
                work_days=hours["work_days"],
                closing_status=closing_status,
            )
        )

    return summaries


@router.get("/{year_month}", response_model=List[PayrollRecordResponse])
def list_payroll_records(
    year_month: str,
    db: Session = Depends(get_db),
    _admin=Depends(get_current_admin_user),
):
    """List all payroll records for a month (admin only)."""
    return (
        db.query(PayrollRecord)
        .filter(PayrollRecord.year_month == year_month)
        .all()
    )


@router.post("/{year_month}/calculate", response_model=List[PayrollRecordResponse])
def calculate_payroll_for_month(
    year_month: str,
    db: Session = Depends(get_db),
    _admin=Depends(get_current_admin_user),
):
    """
    Calculate (or recalculate) payroll for all active users for a given month.
    Upserts a PayrollRecord per user.
    """
    year, month = _get_year_month_parts(year_month)

    users = db.query(User).filter(User.is_active == True).all()
    results = []

    for user in users:
        records = (
            db.query(WorkRecord)
            .filter(
                WorkRecord.user_id == user.id,
                extract("year", WorkRecord.work_date) == year,
                extract("month", WorkRecord.work_date) == month,
            )
            .all()
        )

        hours = aggregate_monthly_hours(records)
        pay = calculate_payroll(user, hours)

        existing = (
            db.query(PayrollRecord)
            .filter(
                PayrollRecord.year_month == year_month,
                PayrollRecord.user_id == user.id,
            )
            .first()
        )

        if existing:
            # Preserve manual adjustments when recalculating
            existing.regular_hours = hours["regular_hours"]
            existing.overtime_hours = hours["overtime_hours"]
            existing.late_night_hours = hours["late_night_hours"]
            existing.holiday_hours = hours["holiday_hours"]
            existing.total_hours = hours["total_hours"]
            existing.base_salary = pay["base_salary"]
            existing.overtime_pay = pay["overtime_pay"]
            existing.late_night_pay = pay["late_night_pay"]
            existing.holiday_pay = pay["holiday_pay"]
            existing.transportation = pay["transportation"]
            existing.allowances = pay["allowances"]
            # Recalculate gross_pay including any existing deductions
            existing.gross_pay = (
                pay["base_salary"]
                + pay["overtime_pay"]
                + pay["late_night_pay"]
                + pay["holiday_pay"]
                + pay["transportation"]
                + pay["allowances"]
                - existing.deductions
            )
            existing.status = "calculated"
            existing.confirmed_at = None
            existing.confirmed_by = None
            db.flush()
            results.append(existing)
        else:
            pr = PayrollRecord(
                year_month=year_month,
                user_id=user.id,
                regular_hours=hours["regular_hours"],
                overtime_hours=hours["overtime_hours"],
                late_night_hours=hours["late_night_hours"],
                holiday_hours=hours["holiday_hours"],
                total_hours=hours["total_hours"],
                base_salary=pay["base_salary"],
                overtime_pay=pay["overtime_pay"],
                late_night_pay=pay["late_night_pay"],
                holiday_pay=pay["holiday_pay"],
                transportation=pay["transportation"],
                allowances=pay["allowances"],
                deductions=0.0,
                gross_pay=pay["base_salary"]
                + pay["overtime_pay"]
                + pay["late_night_pay"]
                + pay["holiday_pay"]
                + pay["transportation"]
                + pay["allowances"],
            )
            db.add(pr)
            db.flush()
            results.append(pr)

    db.commit()
    # Refresh all records so relationships are up-to-date
    for r in results:
        db.refresh(r)
    return results


@router.get("/{year_month}/{user_id}", response_model=PayrollRecordResponse)
def get_payroll_for_user(
    year_month: str,
    user_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """Get payroll for a specific user. Employees can only view their own."""
    if not current_user.is_admin and current_user.id != user_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN, detail="アクセス権限がありません"
        )

    record = (
        db.query(PayrollRecord)
        .filter(
            PayrollRecord.year_month == year_month,
            PayrollRecord.user_id == user_id,
        )
        .first()
    )
    if not record:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="給与記録が見つかりません"
        )
    return record


@router.put("/{year_month}/{user_id}/adjust", response_model=PayrollRecordResponse)
def adjust_payroll(
    year_month: str,
    user_id: int,
    payload: PayrollAdjustment,
    db: Session = Depends(get_db),
    _admin=Depends(get_current_admin_user),
):
    """Update allowances / deductions for a payroll record (admin only)."""
    record = (
        db.query(PayrollRecord)
        .filter(
            PayrollRecord.year_month == year_month,
            PayrollRecord.user_id == user_id,
        )
        .first()
    )
    if not record:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="給与記録が見つかりません"
        )

    update_data = payload.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(record, key, value)

    # Recalculate gross_pay after adjustments
    record.gross_pay = (
        record.base_salary
        + record.overtime_pay
        + record.late_night_pay
        + record.holiday_pay
        + record.transportation
        + record.allowances
        - record.deductions
    )

    db.commit()
    db.refresh(record)
    return record


@router.post("/{year_month}/{user_id}/confirm", response_model=PayrollRecordResponse)
def confirm_payroll(
    year_month: str,
    user_id: int,
    db: Session = Depends(get_db),
    current_admin=Depends(get_current_admin_user),
):
    """Confirm (lock) a payroll record (admin only)."""
    record = (
        db.query(PayrollRecord)
        .filter(
            PayrollRecord.year_month == year_month,
            PayrollRecord.user_id == user_id,
        )
        .first()
    )
    if not record:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="給与記録が見つかりません"
        )

    record.status = "confirmed"
    record.confirmed_at = datetime.utcnow()
    record.confirmed_by = current_admin.id

    db.commit()
    db.refresh(record)
    return record

File: backend/src/routes/payroll.js
const express = require('express');
const router = express.Router();
const { db } = require('../database');
const { authMiddleware, adminMiddleware } = require('../auth');
const { aggregateMonthlyHours } = require('../services/calculation');
const { calculatePayroll } = require('../services/payroll');

router.get('/:yearMonth/summary', authMiddleware, adminMiddleware, (req, res) => {
  const { yearMonth } = req.params;
  const users = db.prepare('SELECT * FROM users WHERE is_active=1').all();
  const result = users.map(u => {
    const records = db.prepare("SELECT * FROM work_records WHERE user_id=? AND work_date LIKE ?").all(u.id, `${yearMonth}-%`);
    const hours = aggregateMonthlyHours(records);
    const closing = db.prepare('SELECT status FROM monthly_closings WHERE year_month=? AND user_id=?').get(yearMonth, u.id);
    return { user_id: u.id, user_name: u.name, employee_id: u.employee_id, year_month: yearMonth, ...hours, closing_status: closing?.status || 'open' };
  });
  res.json(result);
});

router.get('/:yearMonth', authMiddleware, adminMiddleware, (req, res) => {
  const records = db.prepare(`SELECT pr.*, u.name as user_name, u.employee_id FROM payroll_records pr JOIN users u ON pr.user_id=u.id WHERE pr.year_month=? AND u.is_admin=0 ORDER BY u.employee_id`).all(req.params.yearMonth);
  res.json(records);
});

router.post('/:yearMonth/calculate', authMiddleware, adminMiddleware, (req, res) => {
  try {
    const { yearMonth } = req.params;
    const users = db.prepare('SELECT * FROM users WHERE is_active=1 AND is_admin=0').all();
    const upsert = db.prepare(`INSERT INTO payroll_records (year_month,user_id,regular_hours,overtime_hours,late_night_hours,holiday_hours,total_hours,base_salary,overtime_pay,late_night_pay,holiday_pay,transportation,allowances,deductions,gross_pay,status) VALUES (@year_month,@user_id,@regular_hours,@overtime_hours,@late_night_hours,@holiday_hours,@total_hours,@base_salary,@overtime_pay,@late_night_pay,@holiday_pay,@transportation,@allowances,@deductions,@gross_pay,'calculated') ON CONFLICT(year_month,user_id) DO UPDATE SET regular_hours=excluded.regular_hours,overtime_hours=excluded.overtime_hours,late_night_hours=excluded.late_night_hours,holiday_hours=excluded.holiday_hours,total_hours=excluded.total_hours,base_salary=excluded.base_salary,overtime_pay=excluded.overtime_pay,late_night_pay=excluded.late_night_pay,holiday_pay=excluded.holiday_pay,transportation=excluded.transportation,allowances=excluded.allowances,gross_pay=excluded.gross_pay,status='calculated',updated_at=datetime('now','localtime')`);
    const tx = db.transaction(() => {
      for (const u of users) {
        const records = db.prepare("SELECT * FROM work_records WHERE user_id=? AND work_date LIKE ?").all(u.id, `${yearMonth}-%`);
        const hours = aggregateMonthlyHours(records);
        const pay = calculatePayroll(u, hours);
        upsert.run({ year_month: yearMonth, user_id: u.id, ...hours, ...pay });
      }
    });
    tx();
    const result = db.prepare(`SELECT pr.*, u.name as user_name FROM payroll_records pr JOIN users u ON pr.user_id=u.id WHERE pr.year_month=? AND u.is_admin=0 ORDER BY u.employee_id`).all(yearMonth);
    res.json(result);
  } catch (err) {
    console.error('給与計算エラー:', err);
    res.status(500).json({ detail: '給与計算中にエラーが発生しました: ' + err.message });
  }
});

router.get('/:yearMonth/:userId', authMiddleware, (req, res) => {
  const { yearMonth, userId } = req.params;
  if (!req.user.is_admin && req.user.id !== parseInt(userId)) return res.status(403).json({ detail: '権限がありません' });
  const r = db.prepare('SELECT * FROM payroll_records WHERE year_month=? AND user_id=?').get(yearMonth, userId);
  if (!r) return res.status(404).json({ detail: '給与データがありません' });
  res.json(r);
});

router.put('/:yearMonth/:userId/adjust', authMiddleware, adminMiddleware, (req, res) => {
  const { yearMonth, userId } = req.params;
  const r = db.prepare('SELECT * FROM payroll_records WHERE year_month=? AND user_id=?').get(yearMonth, userId);
  if (!r) return res.status(404).json({ detail: '給与データがありません' });
  const { allowances, deductions, additional_notes, transportation } = req.body;
  const newAllowances = allowances ?? r.allowances;
  const newDeductions = deductions ?? r.deductions;
  const newTransportation = transportation ?? r.transportation;
  const grossPay = r.base_salary + r.overtime_pay + r.late_night_pay + r.holiday_pay + newTransportation + newAllowances - newDeductions;
  db.prepare(`UPDATE payroll_records SET allowances=?,deductions=?,transportation=?,additional_notes=?,gross_pay=?,updated_at=datetime('now','localtime') WHERE year_month=? AND user_id=?`)
    .run(newAllowances, newDeductions, newTransportation, additional_notes ?? r.additional_notes, grossPay, yearMonth, userId);
  res.json(db.prepare('SELECT * FROM payroll_records WHERE year_month=? AND user_id=?').get(yearMonth, userId));
});

router.post('/:yearMonth/:userId/confirm', authMiddleware, adminMiddleware, (req, res) => {
  const { yearMonth, userId } = req.params;
  db.prepare(`UPDATE payroll_records SET status='confirmed',confirmed_at=datetime('now','localtime'),confirmed_by=? WHERE year_month=? AND user_id=?`).run(req.user.id, yearMonth, userId);
  res.json(db.prepare('SELECT * FROM payroll_records WHERE year_month=? AND user_id=?').get(yearMonth, userId));
});

module.exports = router;
